source/heating_model/shear_heating.cc

```diff
@@ -76,6 +78,75 @@ namespace aspect
+              AssertThrow(elastic_out != nullptr, ExcMessage("ElasticAdditionalOutputs are requested, but have not been computed."));
+
+              const double shear_modulus = elastic_out->elastic_shear_moduli[q];
+
+              // Retrieve the elastic timestep and viscosity, only two material models
+              // support elasticity.
+              double elastic_timestep = 0.;
+              double elastic_viscosity = 0.;
+              if (Plugins::plugin_type_matches<MaterialModel::ViscoPlastic<dim>>(this->get_material_model()))
+                {
+                  const MaterialModel::ViscoPlastic<dim> &vp = Plugins::get_plugin_as_type<const MaterialModel::ViscoPlastic<dim>>(this->get_material_model());
+                  elastic_viscosity = vp.get_elastic_viscosity(shear_modulus);
+                  elastic_timestep = vp.get_elastic_timestep();
+                }
+              else if (Plugins::plugin_type_matches<MaterialModel::Viscoelastic<dim>>(this->get_material_model()))
+                {
+                  const MaterialModel::Viscoelastic<dim> &ve = Plugins::get_plugin_as_type<const MaterialModel::Viscoelastic<dim>>(this->get_material_model());
+                  elastic_viscosity = ve.get_elastic_viscosity(shear_modulus);
+                  elastic_timestep = ve.get_elastic_timestep();
+                }
+              else
+                AssertThrow(false, ExcMessage("Shear heating cannot be used with elasticity for material models other than ViscoPlastic and Viscoelastic."));
+
+              double dtc = this->get_timestep();
+              if (dtc == 0 && this->get_timestep_number() == 0)
+                dtc = std::min(std::min(this->get_parameters().maximum_time_step, this->get_parameters().maximum_first_time_step), elastic_timestep);
+              const double timestep_ratio = dtc / elastic_timestep;
+              // Scale the elastic viscosity with the timestep ratio, eta is already scaled.
+              elastic_viscosity *= timestep_ratio;
+
+              // Apply the stress update to get the total stress of timestep t.
+              stress = 2. * material_model_outputs.viscosities[q] * deviatoric_strain_rate + material_model_outputs.viscosities[q] / elastic_viscosity * stress_0 +
+                       (1. - timestep_ratio) * (1. - material_model_outputs.viscosities[q] / elastic_viscosity) * stress_old;
+
+              SymmetricTensor<2, dim> visco_plastic_strain_rate = material_model_inputs.strain_rate[q] - ((stress - stress_0) / (2. * dtc * shear_modulus));
+
+              if (this->get_material_model().is_compressible())
+                visco_plastic_strain_rate = visco_plastic_strain_rate -
+                                            1. / 3. * trace(visco_plastic_strain_rate) * unit_symmetric_tensor<dim>();
+
+              heating_model_outputs.heating_source_terms[q] = stress * visco_plastic_strain_rate;
+            }
+
```

Comment:
It is not really duplicating calculations in the material model, but this assembly of the updated stress tensor is done in several plugins outside of the material model (here and in the visu postprocessors). I can imagine creating a function in the material model that returns the fully updated stress tensor of the current timestep, so that does lines 85-140 of shear_heating.cc.

Reply:
Hmmm. I still think it might be better to put the bulk of this calculation in the material model, because:
- A lot of this work is already being done in `Elasticity<dim>::fill_elastic_force_outputs` and `Elasticity<dim>::fill_reaction_outputs`.
- A different formulation of the elasticity problem (a more material-oriented model, for example) might require a different formulation for the heating term.
- The elastic strain rate tensor is something we might want to use elsewhere (and/or visualise directly). For example, we might want to decompose the various components of the strain.

Could a viscous_dissipation term be computed in `elasticity.cc` and passed to an object in `ElasticOutputs` or `ElasticAdditionalOutputs`? Then `shear_heating.cc` could use that object:
`if (this->get_parameters().enable_elasticity == true)`
`   heating_model_outputs.heating_source_terms[q] = elastic_out->viscous_dissipation[i]`

I might be missing something that would make this hard to do.